File: Bubble_Sort/Javascript/madisonmay/bubble_sort.js
//
// Bubble Sort
// -----------
// Advantages:
//     - stable
//     - in-place
// Disadvantages:
//     - slow for large data sets
// Time Complexity:
//     - worst:   O(n^2)
//     - average: O(n^2)
//     - best:    O(n)
// Space Complexity:
//     - O(1)
//

function bubble_sort(l) {
    // basic optimization to avoid unneeded checks
    // note that bubble_sort is the generic bad sorting algorithm
    var copy = l;
    var size = copy.length;
    var sorted_list = false;
    var j = 0;
    while (!sorted_list) {
        var modified = false;
        for (var i = 0; i < (size - j); i++) {
            if (copy[i-1] > copy[i]) {
                var temp = copy[i-1];
                copy[i-1] = copy[i];
                copy[i] = temp;
                modified = true;
            }
        }
        if (!modified) {
            sorted_list = true;
            j += 1;
        }
    }
    return copy;
}

// Test cases
var l = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1];
console.log('Unsorted list: ', l);
l = bubble_sort(l);
console.log('Sorted list: ', l);

File: Bubble_Sort/Python/madisonmay/bubble_sort.py
def bubble_sort(l):
    """
    Bubble Sort
    -----------
    Advantages:
        - stable
        - in-place
    Disadvantages:
        - slow for large data sets
    Time Complexity:
        - worst:   O(n^2)
        - average: O(n^2)
        - best:    O(n)
    Space Complexity:
        - O(1)
    """
    temp = l[:]
    size = len(temp)
    sorted_list = False;
    while not sorted_list:
        modified = False;
        for i in range(1, size):
            if temp[i-1] > temp[i]:
                modified = True;
                temp[i], temp[i-1] = temp[i-1], temp[i]
        if not modified:
            sorted_list = True
    return temp

if __name__ == '__main__':
    # test cases
    help(bubble_sort)
    l = list(range(10, 0, -1))
    print('Unsorted list: ' + str(l));
    l = bubble_sort(l)
    print('Sorted list: ' + str(l))
